### ISSUE TITLE
Intersection of circles on surface

### ISSUE THREAD

**Issue by sammko** [mrJean1/PyGeodesy#41]
Is there any way to get the intersections of two circles given by the center LatLon and radius? There is the trilaterate functions which is similar, could it be (ab)used in some way to get the result?
Thank you.

**Comment by mrJean1** [comment]
Thank you for asking.  Will look into that.

**Comment by sammko** [comment]
I threw together the following according to this SO answer https://stackoverflow.com/a/53331953/2241930:

```python
def circ_intersect(s1, r1, s2, r2):
    R = 6371008.77141
    lat1 = radians(s1.lat)
    lon1 = radians(s1.lon)
    lat2 = radians(s2.lat)
    lon2 = radians(s2.lon)
    B = formy.bearing_(lat1, lon1, lat2, lon2)
    D = formy.haversine_(lat2, lat1, lon2-lon1)
    a1 = r1/R
    a2 = r2/R
    tx = (cos(a2)/cos(a1)-cos(D))/sin(D)
    try:
        x = acos(tx/tan(a1))
    except ValueError:
        return ()
    def ll(br):
        lat = asin(sin(lat1)*cos(a1) + cos(lat1)*sin(a1)*cos(br))
        lon = lon1 + atan2(sin(br)*sin(a1)*cos(lat1), cos(a1)-sin(lat1)*sin(lat2))
        return degrees(lat), degrees(lon)
    return snv.LatLon(*ll(B+x)), snv.LatLon(*ll(B-x))
```

Keep in mind, however, that I have no idea what I am doing and don't know the details and gotchas and whatnot. The results seem reasonable enough, but they don't match up perfectly with the circles drawn by the `folium` library. Of course I don't know how precise those are.

**Comment by mrJean1** [comment]
Will look into this and by back soon.  Also, for accuracy, we may need a separate version for spherical and ellipsoidal models.

**Comment by sammko** [comment]
fwiw, after replacing the final trig calculations with `LatLon.destination`, the precision improved dramatically. The nice thing is that it can be easily verified by measuring the distances from the calculated points to the centers of the original circles, and my results are in the order of micrometers.
```python
def circ_intersect(s1, r1, s2, r2):
    R = 6371008.77141
    lat1 = radians(s1.lat)
    lon1 = radians(s1.lon)
    lat2 = radians(s2.lat)
    lon2 = radians(s2.lon)
    B = radians(s1.initialBearingTo(s2))
    D = s1.distanceTo(s2)/R
    a1 = r1/R
    a2 = r2/R
    tx = (cos(a2)/cos(a1)-cos(D))/sin(D)
    try:
        x = acos(tx/tan(a1))
    except ValueError:
        return ()
    return (s1.destination(r1, degrees(B+x)), s1.destination(r1, degrees(B-x)))
```

**Comment by mrJean1** [comment]
This looks better indeed.  Also, Nvector-based earth models may be a good fit for this case.  Being investigated.

/Jean

**Comment by sammko** [comment]
Here is a testing harness I slapped together https://gist.github.com/sammko/3bce10656778e670629c2b63c62f7357. The output of the second to last cell is the maximum measured deviation, and it hovers around a couple millimetres. The last cell draws a map, altough the circles are not even close to precise. (It doesn't work in the gist, you can click the colab badge at the top of the gist to open in google colab)

**Comment by mrJean1** [comment]
Thank you, I have the spherical intersections code ready and will run your test.  Once the tests pass, I’ll release another PyGeodesy version.  The elliptical case will be something a later release.

**Comment by mrJean1** [comment]
[PyGeodesy 20.6.12](https://github.com/mrJean1/PyGeodesy) includes 1 new function `sphericalTrigonometry.intersections2` and 2 new methods `sphericalTrigonometry.LatLon.intersections2` and `spherical Cartesian.intersections2` to compute the intersection points of 2 circles.  All return a 2-tuple with the intersection point(s) or raise an `IntersectionError` for non-intersecting circles or other issues.

The `sphericalTrigonometry.intersections2` function is (a variation of) your `circ_intersect` function, the `Cartesian.intersections2` method is based on ECEF coordinates.  Also, the `testSpherical` module includes a number of test cases based on your harness.

Thank you very much for all your help.  Don't hesitate to report any issues, otherwise please close this issue.

**Comment by sammko** [comment]
No worries, glad to have been of help.

**Comment by mrJean1** [comment]
My apologies.  There was a `print` statement left in the `Cartesian` `intersections2` method.  [PyGeodesy 20.6.14](https://github.com/mrJean1/PyGeodesy) removes that and adds internal function `_rads3` shared between that method and the `sphericalTrigonometry.intersections2` version.  Discard 20.6.12.

**Comment by mrJean1** [comment]
Sorry again.  There is another issue in `Cartesian` `intersections2` only.  The returned intersections are incorrect if `rad1` is less than `rad2` in `radians`.

That has been fixed in [PyGeodesy 20.6.15](https://github.com/mrJean1/PyGeodesy).  Please discard 20.6.14 and 20.6.12.

**Comment by mrJean1** [comment]
Function `sphericalTrigonometry.intersections2` did not handle _internally_ abutting circles quite correctly.  That has been fixed in upcoming [Pygeodesy 20.7.29](https://pypi.python.org/pypi/PyGeodesy/).  Like for _externally_ abutting circles, the result in that case is a 2-tuple of twice the same instance.

**Comment by mrJean1** [comment]
Like `sphericalTrigonometry.intersections2` there are now similar functions `intersections2` in _ellipsoidal_ modules `ellipsoidalExact`, `-Karney` and `-Vincenty`.  Plus a method called `intersections2` for the _ellipsoidal_ `LatLon` classes.
